### PR TITLE
Added match override logic

### PR DESCRIPTION
**Pull request by tlulu**

### CODE REVIEW

File: src/main/scala/com/shuttleql/services/game/GameServiceServlet.scala [hunk 1]
```diff
@@ -80,4 +82,16 @@ class GameServiceServlet extends GameServiceStack with JacksonJsonSupport {
...
   }
 
+  put("/override") {
+    val matchOverride = parsedBody.extract[MatchOverrideParams]
+    val userId = matchOverride.userId
+    val userId2 = matchOverride.userId2
+
+    MatchMaker.swap(userId, userId2) match {
+      case true =>
+        Ok(JObject(obj = List()))
```
Comment (jasonf7) [inline src/main/scala/com/shuttleql/services/game/GameServiceServlet.scala]: `NoContent(reason = "Success")`

File: src/main/scala/com/shuttleql/services/game/matchmaking/MatchMaker.scala [hunk 2]
```diff
@@ -75,6 +75,121 @@ object MatchMaker {
...
+        },
+        courtName = match2.courtName,
+        courtId = match2.courtId,
+        courtType = match2.courtType
+      )
+
+      if (match2Index < match1Index) {
+        matches = (split1 :+ newMatch2) ++ (split2 :+ newMatch1) ++ split3
+      }
+      else {
```
Comment (jasonf7) [inline src/main/scala/com/shuttleql/services/game/matchmaking/MatchMaker.scala]: `} else {`

File: src/main/scala/com/shuttleql/services/game/matchmaking/MatchMaker.scala [hunk 3]
```diff
@@ -75,6 +75,121 @@ object MatchMaker {
...
+      },
+      courtName = court.courtName,
+      courtId = court.courtId,
+      courtType = court.courtType
+    )
+
+    matches = (split1 :+ newMatch) ++ split2
+    var newPlayerQ: mutable.Queue[Player] = mutable.Queue()
+    newPlayerQ.enqueue(player1)
+    playerQ.filter(x => x.id != userId2).foreach((i: Player) => newPlayerQ.enqueue(i))
```
Comment (jasonf7) [inline src/main/scala/com/shuttleql/services/game/matchmaking/MatchMaker.scala]: `newPlayerQ ++= playerQ.filter(...)`

File: src/main/scala/com/shuttleql/services/game/matchmaking/MatchMaker.scala [hunk 4]
```diff
@@ -75,6 +75,121 @@ object MatchMaker {
...
+      )
+
+      if (match2Index < match1Index) {
+        matches = (split1 :+ newMatch2) ++ (split2 :+ newMatch1) ++ split3
+      }
+      else {
+        matches = (split1 :+ newMatch1) ++ (split2 :+ newMatch2) ++ split3
+      }
+    }
+    else if (match1Index != -1 && match2Index == -1) {
```
Comment (jasonf7) [inline src/main/scala/com/shuttleql/services/game/matchmaking/MatchMaker.scala]: `} else if {`

File: src/main/scala/com/shuttleql/services/game/matchmaking/MatchMaker.scala [hunk 5]
```diff
@@ -75,6 +75,121 @@ object MatchMaker {
...
+      // 2. Player 1 is in a match, player 2 is in the queue
+      swapPlayerInMatchAndQueue(userId, userId2, match1Index)
+    }
+    else if (match1Index == -1 && match2Index != -1) {
```
Comment (jasonf7) [inline src/main/scala/com/shuttleql/services/game/matchmaking/MatchMaker.scala]: `} else if {`

File: src/main/scala/com/shuttleql/services/game/GameServiceServlet.scala [hunk 6]
```diff
@@ -6,10 +6,12 @@ import org.json4s.JsonAST.JObject
 import org.json4s.ext.EnumNameSerializer
 import org.scalatra.json._
 import org.json4s.{DefaultFormats, Formats}
-import org.scalatra.{BadRequest, Ok}
+import org.scalatra._
 import com.typesafe.config._
 import com.gandalf.HMACAuth
 
+case class MatchOverrideParams(userId: Int, userId2: Int)
```
Comment (jasonf7) [inline src/main/scala/com/shuttleql/services/game/GameServiceServlet.scala]: move to `package.scala`

File: src/main/scala/com/shuttleql/services/game/matchmaking/MatchMaker.scala [hunk 7]
```diff
@@ -75,6 +75,121 @@ object MatchMaker {
...
+
+    // 1. Both players are in a match
+    if (match1Index != -1 && match2Index != -1) {
+      var splitIndex = match1Index
+      var splitIndex2 = match2Index
+
+      if (match2Index < match1Index) {
+        splitIndex = match2Index
+        splitIndex2 = match1Index
+      }
```
Comment (jasonf7) [inline src/main/scala/com/shuttleql/services/game/matchmaking/MatchMaker.scala]: `
splitIndex = Math.min(...)
splitIndex2 = Math.max(...)
`

File: src/main/scala/com/shuttleql/services/game/matchmaking/MatchMaker.scala [hunk 8]
```diff
@@ -75,6 +75,121 @@ object MatchMaker {
...
+
+      var split1: List[Match] = List()
+      var split2: List[Match] = List()
+      var split3: List[Match] = List()
+
+      if (splitIndex != 0) {
+        split1 = matches.slice(0, splitIndex)
```
Comment (jasonf7) [inline src/main/scala/com/shuttleql/services/game/matchmaking/MatchMaker.scala]: `val split1 = if (splitIndex != 0) matches.slice(...) else List()`

File: src/main/scala/com/shuttleql/services/game/matchmaking/MatchMaker.scala [hunk 9]
```diff
@@ -75,6 +75,121 @@ object MatchMaker {
...
+        team2 = match2.team2.map { player =>
+          if (player == player2) player1 else player
+        },
+        courtName = match2.courtName,
+        courtId = match2.courtId,
+        courtType = match2.courtType
+      )
+
+      if (match2Index < match1Index) {
+        matches = (split1 :+ newMatch2) ++ (split2 :+ newMatch1) ++ split3
```
Comment (jasonf7) [inline src/main/scala/com/shuttleql/services/game/matchmaking/MatchMaker.scala]: add 

```
val firstMatch = if (match2Index < match1Index) newMatch2 else newMatch1
val secondMatch = ...
```

**Review by jasonf7 — CHANGES_REQUESTED** [review]

File: src/main/scala/com/shuttleql/services/game/GameServiceServlet.scala (hunk 6 above)
Comment (tlulu) [reply src/main/scala/com/shuttleql/services/game/GameServiceServlet.scala]: aight

File: src/main/scala/com/shuttleql/services/game/GameServiceServlet.scala (hunk 1 above)
Comment (tlulu) [reply src/main/scala/com/shuttleql/services/game/GameServiceServlet.scala]: k

File: src/main/scala/com/shuttleql/services/game/matchmaking/MatchMaker.scala (hunk 7 above)
Comment (tlulu) [reply src/main/scala/com/shuttleql/services/game/matchmaking/MatchMaker.scala]: rip

File: src/main/scala/com/shuttleql/services/game/matchmaking/MatchMaker.scala (hunk 8 above)
Comment (tlulu) [reply src/main/scala/com/shuttleql/services/game/matchmaking/MatchMaker.scala]: omg

File: src/main/scala/com/shuttleql/services/game/matchmaking/MatchMaker.scala (hunk 2 above)
Comment (tlulu) [reply src/main/scala/com/shuttleql/services/game/matchmaking/MatchMaker.scala]: fine

File: src/main/scala/com/shuttleql/services/game/matchmaking/MatchMaker.scala (hunk 4 above)
Comment (tlulu) [reply src/main/scala/com/shuttleql/services/game/matchmaking/MatchMaker.scala]: 👿

File: src/main/scala/com/shuttleql/services/game/matchmaking/MatchMaker.scala (hunk 5 above)
Comment (tlulu) [reply src/main/scala/com/shuttleql/services/game/matchmaking/MatchMaker.scala]: 🐙

File: src/main/scala/com/shuttleql/services/game/matchmaking/MatchMaker.scala (hunk 3 above)
Comment (tlulu) [reply src/main/scala/com/shuttleql/services/game/matchmaking/MatchMaker.scala]: oh nice

File: src/main/scala/com/shuttleql/services/game/matchmaking/MatchMaker.scala (hunk 9 above)
Comment (tlulu) [reply src/main/scala/com/shuttleql/services/game/matchmaking/MatchMaker.scala]: Nice

File: src/main/scala/com/shuttleql/services/game/matchmaking/MatchMaker.scala [hunk 10]
```diff
@@ -75,6 +75,102 @@ object MatchMaker {
...
+
+    return true
+  }
+
+  def swapPlayerInMatchAndQueue(userId: Int, userId2: Int, matchIndex: Int) {
+    var split1: List[Match] = List()
+    var split2: List[Match] = List()
+    if (matchIndex != 0) {
+      split1 = matches.slice(0, matchIndex)
+    }
```
Comment (jasonf7) [inline src/main/scala/com/shuttleql/services/game/matchmaking/MatchMaker.scala]: same

**Review by jasonf7 — APPROVED** [review]

**Pull request merged**